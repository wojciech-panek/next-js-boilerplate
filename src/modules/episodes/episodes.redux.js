import { createActions, createReducer } from 'reduxsauce';
import { Record, List, fromJS, Map } from 'immutable';


export const { Types: EpisodesTypes, Creators: EpisodesActions } = createActions({
  fetch: [],
  fetchSuccess: ['data'],
  fetchError: ['payload'],
  fetchSingle: ['id'],
  fetchSingleSuccess: ['data'],
  fetchSingleError: ['payload'],
}, { prefix: 'EPISODES_' });

const EpisodesRecord = new Record({
  items: List(),
  data: Map(),
});

export const INITIAL_STATE = new EpisodesRecord({});

const getSuccessHandler = (state = INITIAL_STATE, action) => state.set('items', fromJS(action.data));

const getSingleSuccessHandler = (state = INITIAL_STATE, action) => state.set('data', fromJS(action.data));

export const reducer = createReducer(INITIAL_STATE, {
  [EpisodesTypes.FETCH_SUCCESS]: getSuccessHandler,
  [EpisodesTypes.FETCH_SINGLE_SUCCESS]: getSingleSuccessHandler,
});
