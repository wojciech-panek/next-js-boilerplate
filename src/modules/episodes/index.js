export {
  EpisodesTypes,
  EpisodesActions,
} from './episodes.redux';

export * from './episodes.selectors';
