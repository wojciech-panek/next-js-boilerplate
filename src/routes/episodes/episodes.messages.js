/* eslint-disable max-len */
import { defineMessages } from 'react-intl';


export default defineMessages({
  pageTitle: {
    id: 'episodes.pageTitle',
    defaultMessage: 'Episode',
  },
  back: {
    id: 'episodes.back',
    defaultMessage: 'Back',
  },
});
