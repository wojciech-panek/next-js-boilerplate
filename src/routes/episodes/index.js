export { default as Episodes } from './episodes.container';
