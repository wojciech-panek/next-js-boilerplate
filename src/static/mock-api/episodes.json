[
  {
    "id": 1,
    "series": 1,
    "number": "01",
    "title": "DII1 Episode 1 Title"
  },
  {
    "id": 2,
    "series": 1,
    "number": "02",
    "title": "DII1 Episode 2 Title"
  },
  {
    "id": 3,
    "series": 1,
    "number": "03",
    "title": "DII1 Episode 3 Title"
  },
  {
    "id": 4,
    "series": 2,
    "number": "01",
    "title": "DII2 Episode 1 Title"
  },
  {
    "id": 5,
    "series": 2,
    "number": "02",
    "title": "DII2 Episode 2 Title"
  },
  {
    "id": 6,
    "series": 2,
    "number": "03",
    "title": "DII2 Episode 3 Title"
  },
  {
    "id": 7,
    "series": 3,
    "number": "01",
    "title": "Liars Episode 1 Title"
  },
  {
    "id": 8,
    "series": 3,
    "number": "02",
    "title": "Liars Episode 2 Title"
  },
  {
    "id": 9,
    "series": 3,
    "number": "03",
    "title": "Liars Episode 3 Title"
  },
  {
    "id": 10,
    "series": 4,
    "number": "01",
    "title": "FML Episode 1 Title"
  },
  {
    "id": 11,
    "series": 4,
    "number": "02",
    "title": "FML Episode 2 Title"
  },
  {
    "id": 12,
    "series": 4,
    "number": "03",
    "title": "FML Episode 3 Title"
  }
]
