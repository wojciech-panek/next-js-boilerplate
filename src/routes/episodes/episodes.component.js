import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import { FormattedMessage } from 'react-intl';
import Head from 'next/head';

import { Loader } from '../../shared/components/loader';
import { Link } from '../../shared/components/link';
import { H2 } from '../../shared/theme/typography';
import { Container } from './episodes.styles';
import messages from './episodes.messages';


export class Episodes extends PureComponent {
  static propTypes = {
    data: PropTypes.object,
    router: PropTypes.object.isRequired,
  };

  getData = () => {
    if (this.props.data.size > 0 && this.props.data.get('id') === parseInt(this.props.router.query.id, 10)) {
      return this.props.data;
    }
    return null;
  };

  render() {
    const data = this.getData();

    if (!data) {
      return <Loader />;
    }

    return (
      <Container>
        <FormattedMessage {...messages.pageTitle}>
          {pageTitle => (
            <Head>
              <title key="title">{ pageTitle }</title>
            </Head>
          )}
        </FormattedMessage>

        <H2>{data.get('title')}</H2>

        <Link to={`/series/${data.get('series')}`}>
          <a><FormattedMessage {...messages.back} /></a>
        </Link>
      </Container>
    );
  }
}
