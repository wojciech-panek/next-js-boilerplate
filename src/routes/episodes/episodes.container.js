import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import { createStructuredSelector } from 'reselect';
import { compose } from 'ramda';
import { withRouter } from 'next/router';

import withIntl from '../../shared/lib/withIntl';
import { selectEpisodesData } from '../../modules/episodes';
import { Episodes } from './episodes.component';


const mapStateToProps = createStructuredSelector({
  data: selectEpisodesData,
});

export const mapDispatchToProps = (dispatch) => bindActionCreators({}, dispatch);

export default compose(
  withIntl,
  withRouter,
  connect(mapStateToProps, mapDispatchToProps),
)(Episodes);
