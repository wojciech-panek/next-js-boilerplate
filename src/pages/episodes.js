import React, { PureComponent } from 'react';
import { compose } from 'ramda';
import { withRouter } from 'next/router';

import withIntl from '../shared/lib/withIntl';
import { Episodes } from '../routes/episodes';
import { EpisodesActions } from '../modules/episodes';


export class EpisodesPage extends PureComponent {
  static fetchData(dispatch, query) {
    dispatch(EpisodesActions.fetchSingle(parseInt(query.id, 10)));
  }

  static async getInitialProps({ isServer, store, query }) {
    if (isServer) {
      await store.execSagaTasks(isServer, (dispatch) => {
        EpisodesPage.fetchData(dispatch, query);
      });
    } else {
      EpisodesPage.fetchData(store.dispatch, query);
    }

    return {};
  }

  render() {
    return (
      <Episodes />
    );
  }
}

export default compose(
  withIntl,
  withRouter,
)(EpisodesPage);
