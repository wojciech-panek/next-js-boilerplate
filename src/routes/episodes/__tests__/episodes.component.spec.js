import React from 'react';
import { shallow } from 'enzyme';
import { fromJS } from 'immutable';

import { Episodes } from '../episodes.component';


describe('Episodes: Component', () => {
  const defaultProps = {
    data: fromJS({ id: 1, series: 1, title: 'example episode title' }),
    router: { query: { id: 1 } },
  };

  const component = (props) => (
    <Episodes {...defaultProps} {...props} />
  );

  const render = (props = {}) => shallow(component(props));

  it('should render correctly when data is provided', () => {
    const wrapper = render();
    global.expect(wrapper).toMatchSnapshot();
  });

  it('should render <Loader /> when data is not provided', () => {
    const wrapper = render({ data: fromJS({}) });
    global.expect(wrapper).toMatchSnapshot();
  });

  it('should render <Loader /> when data does not match', () => {
    const wrapper = render({ router: { query: { id: 2 } } });
    global.expect(wrapper).toMatchSnapshot();
  });
});
