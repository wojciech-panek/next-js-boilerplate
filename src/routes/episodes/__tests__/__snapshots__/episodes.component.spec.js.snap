// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Episodes: Component should render <Loader /> when data does not match 1`] = `<InjectIntl(Loader) />`;

exports[`Episodes: Component should render <Loader /> when data is not provided 1`] = `<InjectIntl(Loader) />`;

exports[`Episodes: Component should render correctly when data is provided 1`] = `
<episodesstyles__Container>
  <FormattedMessage
    defaultMessage="Episode"
    id="episodes.pageTitle"
    values={Object {}}
  >
    <Component />
  </FormattedMessage>
  <typography__H2>
    example episode title
  </typography__H2>
  <InjectIntl(withRouter(Connect(Link)))
    to="/series/1"
  >
    <a>
      <FormattedMessage
        defaultMessage="Back"
        id="episodes.back"
        values={Object {}}
      />
    </a>
  </InjectIntl(withRouter(Connect(Link)))>
</episodesstyles__Container>
`;
